lib/compilers/sail.ts

```diff
@@ -0,0 +1,177 @@
+
+        if (!execOptions.customCwd) {
+            execOptions.customCwd = tmpDir;
+        }
+
+        // We can't use `filters` to decide whether to compile to binary
+        // since it isn't always set (e.g. when this is called from
+        // buildExecutable()). Instead we use a fake command line option.
+        const binary = options.includes("--link-to-binary");
+        options = options.filter(option => option !== "--link-to-binary");
+
+        const fullResult: CompilationResult = {
+            code: 0,
+            timedOut: false,
+            stdout: [],
+            stderr: [],
+            buildsteps: [],
+            inputFilename,
+        };
+
+        const sailResult = await this.doBuildstepAndAddToResult(
+            fullResult,
+            'Sail to C',
+            compiler,
+            [...options, '-o', this.outputFilebase],
+            execOptions,
+        );
+
+        if (sailResult.code !== 0 || !binary) {
+            return fullResult;
+        }
+
+        fullResult.executableFilename = inputFilename + '.exe';
+
+        // Query the sail compiler via `sail -dir` to find out where the
+        // C runtime files are (`sail.c` etc).
+        const sailDirResult = await this.doBuildstepAndAddToResult(
+            fullResult,
+            'Get Sail dir',
+            compiler,
+            ['-dir'],
+            execOptions,
+        );
+
+        if (sailDirResult.code !== 0) {
+            return fullResult;
+        }
+
+        const sailDir = sailDirResult.stdout.map(line => line.text).join('\n').trim();
+
+        // Now compile the C file to an executable.
+        const outputFilenameC = this.getOutputFilename(tmpDir, this.outputFilebase);
+        const outputFilenameExe = outputFilenameC + '.exe';
+
+        const compileResult = await this.doBuildstepAndAddToResult(
+            fullResult,
+            'C to binary',
+            'cc',
+            [
+                outputFilenameC,
+                // Sail C support files.
+                '-I', `${sailDir}/lib`,
+                `${sailDir}/lib/elf.c`,
+                `${sailDir}/lib/rts.c`,
+                `${sailDir}/lib/sail.c`,
+                `${sailDir}/lib/sail_failure.c`,
+                // Support for .elf.gz files. This has been removed in Sail 0.19
+                // so it isn't needed for future versions.
+                '-lz',
+                // For arbitrary precision integer types.
+                '-lgmp',
+                // Enable optimisations so the assembly isn't hilariously verbose.
+                // Ideally this would be user configurable but we'd need
+                // something like `-Wl,..` to pass options through Sail to
+                // here and that doesn't really exist.
+                '-O',
+                '-o', outputFilenameExe,
+            ],
+            execOptions,
+        );
+
+        // This is weird, but as far as I can tell CE expects the same
```

Comment:
Correct; CE expects the file to be called whatever `getOutputFilename` says it should be called.

We've slightly improved this: the `key` or `getOutputFilename` now has a type and you can see which flags are set there and rename accordingly, though it's still a bit tricky to get right.

Reply:
Ah cool, I'll take a look. Are there any other languages that do this that I could use for inspiration?

Reply:
Not really: `ispc` maybe but honestly this is a warty bit of CE code here (as @partouf and I discovered on a pair programming session on `ispc`. Probably stick to what works if it's not instantly obvious how to fix the extension.

Reply:
I changed how the file naming is handled and it seems to make a lot more sense now and still works.

Reply:
(Although it still has this renaming stuff.)

Reply:
Thanks!